src/natives/java_util.ts

```diff
@@ -307,35 +307,126 @@ class java_util_zip_CRC32 {
+  /**
+   * Initialize a new deflater. Using the zlib recommended default values.
+   */
+  public static 'init(IIZ)J'(thread: JVMThread, level: number, strategy: number, nowrap: number): Long {
+    let DEF_MEM_LEVEL = 8; // Zlib recommended default
+    let Z_DEFLATED = 8;    // This value is in the js version of pako under pako.Z_DEFLATED. 
+    // Possibly it is set to private in the Typescript version. The default value is 8, so this should work fine
+
+    let strm = new ZStreamCons();
+    let ret = deflate.deflateInit2(strm, level, Z_DEFLATED, nowrap ? -MAX_WBITS : MAX_WBITS, DEF_MEM_LEVEL, strategy);
+    
+    if (ret != ZlibReturnCode.Z_OK) {
+      let msg = ((strm.msg) ? strm.msg :
+                (ret == ZlibReturnCode.Z_STREAM_ERROR) ?
+                "inflateInit2 returned Z_STREAM_ERROR" :
+                "unknown error initializing zlib library");
+      thread.throwNewException("Ljava/lang/InternalError;", msg);
+    }
+    let num = OpenZStream(strm);
+    return Long.fromNumber(num);
   }
 
+  /**
+   * Apparently this is explicitly not supported by pako.
+   * @see Notes at http://nodeca.github.io/pako/
+   */
   public static 'setDictionary(J[BII)V'(thread: JVMThread, arg0: Long, arg1: JVMTypes.JVMArray<number>, arg2: number, arg3: number): void {
     thread.throwNewException('Ljava/lang/UnsatisfiedLinkError;', 'Native method not implemented.');
   }
 
-  public static 'deflateBytes(J[BIII)I'(thread: JVMThread, javaThis: JVMTypes.java_util_zip_Deflater, arg0: Long, arg1: JVMTypes.JVMArray<number>, arg2: number, arg3: number, arg4: number): number {
-    thread.throwNewException('Ljava/lang/UnsatisfiedLinkError;', 'Native method not implemented.');
+  public static 'deflateBytes(J[BIII)I'(thread: JVMThread, javaThis: JVMTypes.java_util_zip_Deflater, addr: Long, b: JVMTypes.JVMArray<number>, off: number, len: number, flush: number): number {
+    let strm = GetZStream(thread, addr.toNumber());
+    if (!strm) return;
+
+    let thisBuf = javaThis['java/util/zip/Deflater/buf'];
+    let thisOff = javaThis['java/util/zip/Deflater/off'];
+    let thisLen = javaThis['java/util/zip/Deflater/len'];
+
+    let inBuf = thisBuf.array;
+    let outBuf = b.array;
+
+    strm.input = i82u8(inBuf, 0, inBuf.length);
+    strm.next_in = thisOff;
+    strm.avail_in = thisLen;
+
+    strm.output = i82u8(outBuf, 0, outBuf.length);
+    strm.next_out = off;
+    strm.avail_out = len;
+
+    if (javaThis['java/util/zip/Deflater/setParams']) {
+      let level = javaThis['java/util/zip/Deflater/level'];
+      let strategy = javaThis['java/util/zip/Deflater/level'];
+      //deflateParams is not yet supported by pako. We'll open a new ZStream with the new parameters instead.
+      // res = deflate.deflateParams(strm, level, strategy); 
+      let newStream = new ZStreamCons();
+      let res = deflate.deflateInit2(newStream, strm.state.level, strm.state.method, strm.state.windowBits, strm.state.memLevel, strm.state.strategy);
```

Comment:
Should this be `strm.state.level`, or `level`? It looks like you're re-applying the existing stream's settings to the new stream, instead of using the new settings.